plugins/providers/gcloudiam/provider.go

```diff
@@ -99,12 +152,27 @@ func (p *Provider) RevokeAccess(pc *domain.ProviderConfig, a *domain.Appeal) err
 	}
 
 	if a.Resource.Type == ResourceTypeProject || a.Resource.Type == ResourceTypeOrganization {
-		if err := client.RevokeAccess(a.AccountType, a.AccountID, a.Role); err != nil {
-			if errors.Is(err, ErrPermissionNotFound) {
-				return nil
+		resourceRoleMap := make(map[string]*domain.Role)
+		for _, rc := range pc.Resources {
+			for _, ro := range rc.Roles {
+				resourceRole := fmt.Sprintf("%s-%s", rc.Type, ro.ID)
+				resourceRoleMap[resourceRole] = ro
+			}
+		}
+		appealResourceRole := fmt.Sprintf("%s-%s", a.Resource.Type, a.Role)
+		requestedRole, ok := resourceRoleMap[appealResourceRole]
+		if !ok {
+			return ErrInvalidRole
+		}
+		for _, p := range requestedRole.Permissions {
+			permission := fmt.Sprint(p)
+			if err := client.RevokeAccess(a.AccountType, a.AccountID, permission); err != nil {
+				if !errors.Is(err, ErrPermissionNotFound) {
+					return err
+				}
```

Comment:
this logic implemented twice in `GrantAccess` and `RevokeAccess`, can you refactor this by moving the logic into a function and reuse it in those both places.

Reply:
Done